### DISCUSSION TITLE
Update Dogecoin to be Post Quantum Resistant

### DISCUSSION THREAD

**Discussion by cryptaralex** [dogecoin/dogecoin#3400] [General]
The days are numbered for ECC, which is the basis of the signature scheme which Dogecoin has borrowed from its parents. Quantum computers will be able to break ECC within the next decade, possibly sooner.  The cryptographic community understands this risk and has been developing new signature schemes that resist attacks from quantum computers. I propose we use one of the contending schemes from NIST's phase 3 competition  and use that as the basis of a Post Quantum Dogecoin, otherwise known as Dogecoin PQC. Tidecoin is a premined example of a Post Quantum secure cryptocurrency which we could borrow from.

Dogecoin PQC will give Dogecoin a significant market advantage over competing cryptocurrencies.

**Comment by patricklodder** [comment]
> otherwise known as Dogecoin PQC

Got some code for Dogecoin PQC since it is already known?

**Comment by cryptaralex** [comment]
Fork this https://github.com/tidecoin and change the issuance schedule to match dogecoin as a starting point. I haven't done it yet, but it should be fairly trivial. There's a bit more to think through still. How do we make the transition? At what block height do we stop accepting legacy transactions? Tidecoin uses a CPU bound mining algo which is a little better for decentralization, but not really (AMD would love it though). So we want to stick with Scrypt? Or should we move to something absolutely fair like proof of human activity?

**Comment by patricklodder** [comment]
> So we want to stick with Scrypt? Or should we move to something absolutely fair like proof of human activity?

One change at a time. You will want to only propose a new signature algo and a way to migrate to it (send from secp256k1 to that) and keep everything else 100% intact.

We can have some real fun with a custom signet once 1.21 is ready 😁

**Comment by cryptaralex** [comment]
Sounds good. We will start with porting falcon into dogecoin and coming up with a method to send from secp256k1.

**Comment by delbonis** [comment]
Practically speaking this wouldn't even be that hard.  And could be done with a soft fork.  Just add a CHECKFALCONSIGVERIFY opcode in place of a nop.  That's like, all you'd have to do and it'd be fine.

But note that in the smallest version of [Falcon](https://falcon-sign.info/), pubkeys are 897 bytes and signatures are 666 bytes.  This is over an order of magnitude higher costs than with ECDSA, for relatively little short term benefit.  So that's something to consider, and most users won't use these new addresses for a while.  Quantum-secure signature schemes aren't fully mature yet and it seems that the state of the art is still evolving.

**Comment by cryptaralex** [comment]
The Phase 3 competition [conference](https://csrc.nist.gov/Events/2021/third-pqc-standardization-conference) is literally next week. We should be getting close to getting recommendations for standards going forward. CRYSTALS-Dilithium is another good candidate for use in cryptocurrency. We don't expect a single winner in phase 3, rather we believe that there will be multiple recommendations. 

The size requirements are certainly not trivial, though Segwit does help. We must keep this in consideration as we move forward. Initially I wouldn't expect much more than some experimental use of the new address and perhaps use as accumulation for long term holders. Eventually though, only the new addresses will be used.

**Comment by cryptaralex** [comment]
Here's the NIST round 3 [summary](https://csrc.nist.gov/CSRC/media/Presentations/status-update-on-the-3rd-round/images-media/session-1-moody-nist-round-3-update.pdf). For signatures, Falcon is still promising, but there are some patent concerns. Dilithium may very well end up being the NIST recommended solution. It's still a developing field and any of the alternates could end up becoming the standard if a weakness is found. Never mind the fact that the agency involved is known to introduce crypto they can backdoor.

**Comment by patricklodder** [comment]
> Falcon is still promising, but there are some patent concerns.

Do you have a more elaborate resource that outlines patent issues? We wouldn't want shibes to be confronted with having to pay-as-you-go some corporate entity.

**Comment by cryptaralex** [comment]
Indeed. It looks like we have to grab it from [wayback](https://web.archive.org/web/20200616005235/https://cr.yp.to/patents.html) as Daniel J. Bernstein's site is currently down.

**Comment by patricklodder** [comment]
Thanks! I'm not too worried about these patents, based on their expiry dates. At a glance I'd say that we currently have more time than these will be active over. Would love to know a legal shibe's opinion on this though!

**Comment by Sttarrlord** [comment]
QRL seems to have a great team and they did build a QRLcoin. I think we should analyse it and they are also on Bittrex and they have pretty good suit of interfaces.

**Comment by cryptaralex** [comment]
They are trying to force a square peg into a round hole. QRL uses **stateful** XMSS, which really isn't well suited for the use case of a decentralized cryptocurrency. The overhead of keeping track of the one-time private keys that have been used to generate signatures, in order to prevent their reuse, is untenable.

**Comment by ReverseControl** [comment]
On this matter I am of the following opinion:

I strongly favor Dilithium signatures because the corresponding Encryption/Decryption scheme (KYBER) also made it to the final of NIST's PQC competition so that if we want to use Post-Quantum Private/Public Keys in the future the work needed to be done will be substantially reduced. Also, the mathematics behind these two schemes is identical; so even choosing parameters is a lot easier. Otherwise we would have to, at some point, explore parameter spaces in different mathematical spaces for signatures and enc/dec, and that's just unnecessary.

@cryptaralex @patricklodder @delbonis
